packages/react-components/src/hooks/useAlarms/utils/alarmDataUtils.ts

```diff
@@ -0,0 +1,141 @@
+import {
+  AssetProperty,
+  DescribeAssetModelResponse,
+  DescribeAssetResponse,
+} from '@aws-sdk/client-iotsitewise';
+import { AlarmData, AlarmDataStatus, AlarmRequest } from '../types';
+import {
+  buildFromCompositeModel,
+  isAlarmCompositeModel,
+} from './compositeModelUtils';
+
+/**
+ * This function builds an AlarmData object for each alarm composite model on a SiteWise asset.
+ * The content of AlarmData depends on the AlarmRequest argument.
+ *
+ * @param request is an AlarmRequest that may have an assetId
+ * @param status is the status of the DescribeAssetResponse query
+ * @param assetResponse is the DescribeAssetResponse
+ * @returns a list of AlarmData for all alarms on an asset
+ */
+export const buildFromAssetResponse = ({
+  request,
+  status,
+  assetResponse,
+}: {
+  request: AlarmRequest;
+  status: AlarmDataStatus;
+  assetResponse?: DescribeAssetResponse;
+}): AlarmData[] => {
+  const { assetId, assetCompositeModelId, inputPropertyId } = request;
+
+  if (!assetResponse || !assetId) {
+    return [];
+  }
+
+  if (assetResponse && assetId && assetResponse?.assetId !== assetId) {
+    throw new Error(
+      'AlarmDataFactory.buildFromAssetResponse expects there to be an assetId in the request that matches the DescribeAssetResponse'
+    );
+  }
+
+  // Find all alarms
+  const alarmCompositeModels = assetResponse?.assetCompositeModels?.filter(
+    isAlarmCompositeModel
+  );
+
+  /**
+   * Find inputProperty from asset summary if requested.
+   *
+   * Used as an indicator in useAlarms to filter out alarms
+   * when the inputProperty from IoT Events alarmModels don't match.
+   */
+  let inputPropertyList: AssetProperty[] | undefined;
+  if (inputPropertyId) {
+    const inputProperty = assetResponse?.assetProperties?.find(
+      (property) => property.id === inputPropertyId
+    );
+    if (inputProperty) {
+      inputPropertyList = [inputProperty];
+    }
+  }
```

Comment:
I would encapsulate mutative logic like this in its own function. Having logic like this out in the open is going to be prone to errors by accident.